L1Trigger/L1THGCal/src/be_algorithms/HGCalMulticlusteringImpl.cc

```diff
@@ -95,6 +134,74 @@ void HGCalMulticlusteringImpl::clusterize( const edm::PtrVector<l1t::HGCalCluste
     }
     
 }
+void HGCalMulticlusteringImpl::clusterizeDBSCAN( const edm::PtrVector<l1t::HGCalCluster> & clustersPtrs, 
+                                           l1t::HGCalMulticlusterBxCollection & multiclusters)
+{
 
-
-
+  std::vector<l1t::HGCalMulticluster> multiclustersTmp;
+  l1t::HGCalMulticluster mcluTmp;
+  std::vector<bool> visited(clustersPtrs.size(),false);
+  std::vector<bool> merged (clustersPtrs.size(),false);
+  std::vector<std::vector<int>> neighborList;
+  int iclu = 0, imclu = 0;
+
+  for(edm::PtrVector<l1t::HGCalCluster>::iterator clu = clustersPtrs.begin(); clu != clustersPtrs.end(); ++clu, ++iclu){
+      std::vector<int> neighbors;      
+      
+      if(!visited.at(iclu)){
+	visited.at(iclu)=true;
+	findNeighbor(clustersPtrs, **clu, neighbors);
+	neighborList.push_back(neighbors);
+
+	if(neighborList.at(iclu).size() > minNDbscan_) {
+	  multiclustersTmp.emplace_back( *clu );
+	  merged.at(iclu) = true;
+	  
+	  for(unsigned int neighNo = 0; neighNo < neighborList.at(iclu).size(); neighNo++){
```

Comment:
A  range-based loop would make the code more readable, to avoid having all the `neighborList.at(iclu).at(neighNo)`

Reply:
I spent half a day on this, I don't think range-based loops work here (note that range is not constant). I would be more than happy to try if you have a suggestion.

Reply:
Ah yes, right. In that case maybe you can define an integer at the beginning of the loop:
```cpp
int neighbor = neighborList.at(iclu).at(neighNo);
```
to be used later in the loop.